aws/logs_monitoring/cache.py

```diff
@@ -0,0 +1,421 @@
+        "Add the Datadog Lambda layer to this function to submit enhanced metrics."
+    )
+    DD_SUBMIT_ENHANCED_METRICS = False
+
+_other_chars = r"\w:\-\.\/"
+Sanitize = re.compile(r"[^%s]" % _other_chars, re.UNICODE).sub
+Dedupe = re.compile(r"_+", re.UNICODE).sub
+FixInit = re.compile(r"^[_\d]*", re.UNICODE).sub
+
+
+def sanitize_aws_tag_string(tag, remove_colons=False, remove_leading_digits=True):
+    """Convert characters banned from DD but allowed in AWS tags to underscores"""
+    global Sanitize, Dedupe, FixInit
+
+    # 1. Replace colons with _
+    # 2. Convert to all lowercase unicode string
+    # 3. Convert bad characters to underscores
+    # 4. Dedupe contiguous underscores
+    # 5. Remove initial underscores/digits such that the string
+    #    starts with an alpha char
+    #    FIXME: tag normalization incorrectly supports tags starting
+    #    with a ':', but this behavior should be phased out in future
+    #    as it results in unqueryable data.  See dogweb/#11193
+    # 6. Strip trailing underscores
+
+    if len(tag) == 0:
+        # if tag is empty, nothing to do
+        return tag
+
+    if remove_colons:
+        tag = tag.replace(":", "_")
+    tag = Dedupe("_", Sanitize("_", tag.lower()))
+    if remove_leading_digits:
+        first_char = tag[0]
+        if first_char == "_" or "0" <= first_char <= "9":
+            tag = FixInit("", tag)
+    tag = tag.rstrip("_")
+    return tag
+
+
+def send_forwarder_internal_metrics(name, additional_tags=[]):
+    """Send forwarder's internal metrics to DD"""
+    lambda_stats.distribution(
+        "{}.{}".format(DD_FORWARDER_TELEMETRY_NAMESPACE_PREFIX, name),
+        1,
+        tags=get_forwarder_telemetry_tags() + additional_tags,
+    )
+
+
+def should_fetch_custom_tags():
+    """Checks the env var to determine if the customer has opted-in to fetching custom tags"""
+    return os.environ.get("DD_FETCH_LAMBDA_TAGS", "false").lower() == "true"
+
+
+def get_last_modified_time(s3_file):
+    last_modified_str = s3_file["ResponseMetadata"]["HTTPHeaders"]["last-modified"]
+    last_modified_date = datetime.datetime.strptime(
+        last_modified_str, "%a, %d %b %Y %H:%M:%S %Z"
+    )
+    last_modified_unix_time = int(last_modified_date.strftime("%s"))
+    return last_modified_unix_time
+
+
+class LambdaTagsCache(object):
+
+    CACHE_FILENAME = None
+    CACHE_LOCK_FILENAME = None
+
+    def __init__(self, tags_ttl_seconds=DD_TAGS_CACHE_TTL_SECONDS):
+        self.tags_ttl_seconds = tags_ttl_seconds
+
+        self.tags_by_id = {}
+        self.last_tags_fetch_time = 0
+
+    def write_cache_to_s3(self, data):
+        """Writes tags cache to s3"""
+        try:
+            logger.debug("Trying to write data to s3: {}".format(data))
+            s3_object = s3_client.Object(DD_S3_BUCKET_NAME, self.CACHE_FILENAME)
+            s3_object.put(Body=(bytes(json.dumps(data).encode("UTF-8"))))
```

Comment:
Does the lambda log forwarder cloudformation template need to be updated for this IAM permission?